File: aoc/rust/src/days/two.rs
use crate::aoc_util;

struct Sub {
    horiz: i64,
    depth: i64,
    aim: i64,
}

impl Sub {
    fn r#move(&mut self, instruction: &str, count: i64, part_1: bool) {
        match (instruction, count) {
            ("forward", x) => {
                self.horiz += x;
                if !part_1 {
                    self.depth += self.aim * x;
                }
            }

            ("up", x) => {
                if part_1 {
                    self.depth -= x;
                } else {
                    self.aim -= x;
                }
            }

            ("down", x) => {
                if part_1 {
                    self.depth += x
                } else {
                    self.aim += x
                }
            }
            (a, b) => println!("Unexpected: {} {}", a, b),
        }
    }
}

pub fn part_1(input: &String) -> i64 {
    let mut sub = Sub {
        aim: 0,
        depth: 0,
        horiz: 0,
    };
    for line in aoc_util::to_string_vec(input) {
        let mut split = line.split(" ");
        sub.r#move(
            split.next().unwrap(),
            split.next().unwrap().parse().unwrap(),
            true,
        )
    }

    sub.horiz * sub.depth
}

pub fn part_2(input: &String) -> i64 {
    let mut sub = Sub {
        aim: 0,
        depth: 0,
        horiz: 0,
    };
    for line in aoc_util::to_string_vec(input) {
        let mut split = line.split(" ");
        sub.r#move(
            split.next().unwrap(),
            split.next().unwrap().parse().unwrap(),
            false,
        )
    }

    sub.horiz * sub.depth
}

File: aoc/python/two.py
from __future__ import annotations
from dataclasses import dataclass


@dataclass
class Submarine:
    x: int = 0
    aim: int = 0
    depth: int = 0

    def move(self, instruction: str, p2: bool):
        split = instruction.split(" ")
        match (split[0], int(split[1])):
            case ("forward", x):
                self.x += x
                if p2:
                    self.depth += self.aim * x
            case ("up", x):
                if p2:
                    self.aim -= x

                else:
                    self.depth -= x

            case ("down", x):
                if p2:
                    self.aim += x

                else:
                    self.depth += x

            case (a, b):
                print(f"Unknown instruction {a} {b}")


def part_1(input: str) -> str:
    sub = Submarine()
    for instruction in input.splitlines():
        sub.move(instruction, False)

    return str(sub.x * sub.depth)


def part_2(input: str) -> str:
    sub = Submarine()
    for instruction in input.splitlines():
        sub.move(instruction, True)

    return str(sub.x * sub.depth)
